File: 3d/src/main/java/io/github/xerooup/frame3d/core/Game.java
package io.github.xerooup.frame3d.core;

import io.github.xerooup.frame3d.core.internal.Engine;
import io.github.xerooup.frame3d.graphics.Color;
import io.github.xerooup.frame3d.graphics.RenderContext;

public abstract class Game {
    public abstract void settings(Settings settings);

    public void create() {}
    public void update(float delta) {}
    public void render(RenderContext render) {}

    public static void run(Game game) {
        Settings settings = new Settings();
        game.settings(settings);
        new Engine(settings, game).run();
    }

    public static void stop() {
        Engine.forceStop();
    }

    public static int fps = 0;
    public static float uptime = 0f;
}

File: 2d/src/main/kotlin/io/github/xerooup/frame2d/core/Game.kt
package io.github.xerooup.frame2d.core

import io.github.xerooup.frame2d.core.internal.Engine
import io.github.xerooup.frame2d.graphics.Color
import io.github.xerooup.frame2d.graphics.DrawContext

data class Settings(
    var width: Int = 800,
    var height: Int = 600,
    var title: String = "my very mega super cool great amazing awesome game",
    var decorated: Boolean = true,
    var targetFPS: Int = 60,
    var background: Color = Color.WHITE,
    var icon: String? = null,
    var windowX: Int? = null,
    var windowY: Int? = null
)

abstract class Game {
    abstract fun settings(settings: Settings)
    open fun create() {}
    open fun update(delta: Float) {}
    open fun render(draw: DrawContext) {}

    companion object {
        fun run(game: Game) {
            val settings = Settings()
            game.settings(settings)
            Engine(settings, game).run()
        }
        fun stop() {
            Engine.forceStop()
        }
        var fps: Int = 0
            internal set

        var uptime: Float = 0f
            internal set
    }
}